compiler/rustc_hir_analysis/src/check/compare_impl_item.rs

```diff
@@ -558,7 +558,7 @@ fn compare_asyncness<'tcx>(
     trait_item_span: Option<Span>,
 ) -> Result<(), ErrorGuaranteed> {
     if tcx.asyncness(trait_m.def_id) == hir::IsAsync::Async {
-        match tcx.fn_sig(impl_m.def_id).skip_binder().output().kind() {
+        match tcx.fn_sig(impl_m.def_id).subst_identity().skip_binder().output().kind() {
```

Comment:
I think we should use `skip_binder` here as we compare the impl methods with the trait method without instantiating both.

we can look at the self type as we know that the trait method has `ty::Alias` as the self type so we can also require it from the impl self type before instantiating the binder.